File: src/main/java/com/jameslandrum/bluetoothsmart2/scanner/DeviceScanner.java
/*
  Copyright 2016 James Landrum

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

package com.jameslandrum.bluetoothsmart2.scanner;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.os.Build;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import com.jameslandrum.bluetoothsmart2.ScannerCallback;
import com.jameslandrum.bluetoothsmart2.SmartDevice;
import com.jameslandrum.bluetoothsmart2.SmartDeviceCallback;
import com.jameslandrum.bluetoothsmart2.actionqueue.Identifier;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Interface class that connects to the proper device scanner for Android
 */
@SuppressWarnings({"WeakerAccess", "unused"})
public abstract class DeviceScanner {
    private static final byte[] APPLE_PREFIX =
            new byte[]{0x4C, 0x00};

    @IntDef({SCAN_MODE_LOW_LATENCY, SCAN_MODE_LOW_POWER, SCAN_MODE_NORMAL, SCAN_MODE_PASSIVE})
    @Retention(RetentionPolicy.SOURCE)
    public @interface ScanMode {}
    public static final int SCAN_MODE_PASSIVE =     -1;
    public static final int SCAN_MODE_LOW_POWER =   0;
    public static final int SCAN_MODE_NORMAL =      1;
    public static final int SCAN_MODE_LOW_LATENCY = 2;

    private boolean mEnableDiscovery = false;
    protected static int mScanMode;
    protected static int mScanInterval;
    private static DeviceScanner mInstance;
    protected static final ArrayList<ScannerCallback> mListeners = new ArrayList<>();
    protected static final ArrayList<String> mInvalidDevices = new ArrayList<>();
    protected static final ConcurrentHashMap<String, SmartDevice> mDevices = new ConcurrentHashMap<>();
    protected static final HashSet<Identifier> mIdentifiers = new HashSet<>();

    public <T extends SmartDevice> void forgetDevice(T device) {
        mInvalidDevices.remove(device.getAddress());
        mDevices.remove(device.getAddress());
    }

    public <T extends SmartDevice> void injectDevice(T device) {
        mDevices.remove(device.getAddress());
        mInvalidDevices.remove(device.getAddress());
        mDevices.put(device.getAddress(),device);
    }

    /**
     * Returns an instance of the DeviceScanner
     * @return The relevant device scanner for this device.
     */
    public static DeviceScanner getInstance() {
        if (mInstance != null) return mInstance;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            mInstance = new LollipopDeviceScanner();
        } else {
            mInstance = new KitKatDeviceScanner();
        }
        return mInstance;
    }

    DeviceScanner() {
    }

    public abstract void startScan(@ScanMode int scanMode, int interval);
    public abstract void stopScan();
    public abstract boolean isScanning();

    public void addIdentifier(Identifier identifier) {
        mIdentifiers.add(identifier);
        mInvalidDevices.clear();
    }

    @SuppressWarnings("RedundantIfStatement")
    void processAdvertisement(byte[] data, BluetoothDevice device, List<UUID> uuids, int rssi) {
        if (mInvalidDevices.contains(device.getAddress())) return;

        boolean isBeacon = data[5] == APPLE_PREFIX[0] &&
                            data[6] == APPLE_PREFIX[1];

        if (mDevices.containsKey(device.getAddress())) {
            SmartDevice target = mDevices.get(device.getAddress());
            if (isBeacon) {
                target.onBeacon();
            } else {
                target.onAdvertisement(data,rssi);
            }
        } else if (mEnableDiscovery && !isBeacon) {
            Identifier identifier = null;

            getId: for (Identifier i : mIdentifiers) {
                if (i.getName() != null && !i.getName().equals(device.getName())) continue;
                for (UUID uuid: i.getUuids()) {
                    if (!uuids.contains(uuid)) continue getId;
                }
                if (i.getByteId() != null && !i.getByteId().checkBytes(data)) continue;
                identifier = i;
            }

            if (identifier != null) {
                try {
                    Class<? extends SmartDevice> k = identifier.getDeviceClass();

                    SmartDevice target = injectDevice(k, device);
                    target.onAdvertisement(data,rssi);

                    for (ScannerCallback c : mListeners) {
                        c.onDeviceDiscovered(target);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            } else {
                mInvalidDevices.add(device.getAddress());
            }
        }

    }

    public <T extends SmartDevice> T injectDevice(Class<T> k, BluetoothDevice device) throws IllegalAccessException, InstantiationException {
        if (mDevices.containsKey(device.getAddress())) {
            //noinspection unchecked
            return (T) mDevices.get(device.getAddress());
        }

        T target = k.newInstance();
        target.init(device);
        mDevices.put(device.getAddress(), target);
        return target;
    }

    public <T extends SmartDevice> T injectDevice(Class<T> k, String address) throws IllegalAccessException, InstantiationException
    {
        if (mDevices.containsKey(address)) {
            //noinspection unchecked
            return (T) mDevices.get(address);
        }

        T target = k.newInstance();
        target.init(BluetoothAdapter.getDefaultAdapter().getRemoteDevice(address));
        mDevices.put(address, target);
        return target;
    }

    public void enableDiscovery(boolean enableDiscovery) {
        mEnableDiscovery = enableDiscovery;
    }

    public List<SmartDevice> getAllDevices() {
        return Collections.unmodifiableList(new ArrayList<>(mDevices.values()));
    }

    public void addScanListener(@NonNull ScannerCallback listener) {
        mListeners.add(listener);
    }

    public void removeScanListener(@NonNull ScannerCallback listener) {
        mListeners.remove(listener);
    }

    public SmartDevice getDeviceByMacAddress(String macAddress) {
        return mDevices.get(macAddress);
    }
}

File: src/main/java/com/jameslandrum/bluetoothsmart2/scanner/DeviceScanner.kt
package com.jameslandrum.bluetoothsmart2.scanner

import android.annotation.TargetApi
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.os.Build
import android.support.annotation.IntDef
import android.support.annotation.RequiresApi
import com.jameslandrum.bluetoothsmart2.Identifier
import com.jameslandrum.bluetoothsmart2.SmartDevice
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.rxkotlin.toFlowable
import io.reactivex.rxkotlin.toObservable
import java.lang.ref.WeakReference
import java.lang.reflect.InvocationTargetException
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.collections.HashSet

@TargetApi(Build.VERSION_CODES.KITKAT)
abstract class DeviceScanner {
    protected var scanMode: Int = 0
    protected var scanInterval: Int = 0
    protected val invalidDevices = ArrayList<String>()

    val identifiers = HashSet<Identifier>()
    var discovery = false
    val devices = DeviceSet()

    companion object {
        val APPLE_PREFIX = byteArrayOf(0x4C, 0x00)
        const val SCAN_MODE_PASSIVE = -1
        const val SCAN_MODE_LOW_POWER = 0
        const val SCAN_MODE_NORMAL = 1
        const val SCAN_MODE_LOW_LATENCY = 2
        var _instance : DeviceScanner? = null

        val instance: DeviceScanner
            get() {
                if (_instance != null) return _instance!!
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    _instance = LollipopDeviceScanner()
                } else {
                    _instance = KitKatDeviceScanner()
                }
                return _instance!!
            }

        fun startScan(scanMode: Int, interval: Int) = instance.startScan(scanMode,interval)

        @IntDef(SCAN_MODE_LOW_LATENCY.toLong(), SCAN_MODE_LOW_POWER.toLong(), SCAN_MODE_NORMAL.toLong(), SCAN_MODE_PASSIVE.toLong())
        @Retention(AnnotationRetention.SOURCE)
        annotation class ScanMode
    }

    fun <T : SmartDevice> forgetDevice(device: T) {
        invalidDevices.remove(device.address)
        devices.removeIf { it.address == device.address }
    }

    fun <T : SmartDevice> injectDevice(device: T): T {
        devices.removeIf { it.address == device.address }
        invalidDevices -= device.address
        devices += device
        return device
    }

    abstract fun startScan(@ScanMode scanMode: Int, interval: Int)
    abstract fun stopScan()
    abstract fun isScanning(): Boolean

    fun addIdentifier(identifier: Identifier) {
        identifiers += identifier
        invalidDevices.clear()
    }

    @RequiresApi(Build.VERSION_CODES.ECLAIR)
    internal fun processAdvertisement(data: ByteArray, device: BluetoothDevice, uuids: List<UUID>, rssi: Int) {
        if (invalidDevices.contains(device.address)) return

        val isBeacon = data[5] == APPLE_PREFIX[0] && data[6] == APPLE_PREFIX[1]

        if (devices.any { it.address == device.address }) {
            val target = devices[device.address]
            if (isBeacon) {
                target?.onBeacon()
            } else {
                target?.updateAdvertisement(data, rssi)
            }
        } else if (discovery && !isBeacon) {
            var identifier: Identifier? = null

            getId@ for (i in identifiers) {
                if (i.name != null && i.name != device.name) continue
                for (uuid in i.uuids) {
                    if (!uuids.contains(uuid)) continue@getId
                }
                if (!i.byteId.invoke(data)) continue
                identifier = i
            }

            if (identifier != null) {
                try {
                    val k = identifier.deviceClass
                    val target = injectDevice(k, device)
                    target.updateAdvertisement(data, rssi)
                } catch (e: Exception) {
                    e.printStackTrace()
                }

            } else {
                invalidDevices.add(device.address)
            }
        }

    }

    @RequiresApi(Build.VERSION_CODES.ECLAIR)
    @Throws(IllegalAccessException::class, InstantiationException::class, NoSuchMethodException::class, InvocationTargetException::class)
    fun <T : SmartDevice> injectDevice(k: Class<T>, device: BluetoothDevice): SmartDevice {
        if (devices.available(device.address)) return devices[device.address]!!

        val target = k.getConstructor(BluetoothDevice::class.java).newInstance(device)
        devices += target
        return target
    }

    @RequiresApi(Build.VERSION_CODES.ECLAIR)
    @Throws(IllegalAccessException::class, InstantiationException::class, NoSuchMethodException::class, InvocationTargetException::class)
    fun <T : SmartDevice> injectDevice(k: Class<T>, address: String): SmartDevice {
        return this.injectDevice(k, BluetoothAdapter.getDefaultAdapter().getRemoteDevice(address))
    }

    fun getDeviceByMacAddress(macAddress: String): SmartDevice? {
        return devices.get(macAddress)
    }
}

interface OnDeviceUpdateListener {
    fun onDeviceUpdated(device: SmartDevice)
}

interface ScannerCallback {
    fun <T : SmartDevice> onDeviceDiscovered(device: T)
}

class DeviceSet : HashSet<SmartDevice>() {
    private val listeners = WeakListeners()

    operator fun get(macAddress: String) : SmartDevice? = this.find { it.address == macAddress }
    operator fun get(index: Int) : SmartDevice? = this.elementAt(index)
    fun available(macAddress: String) = this.any { it.address == macAddress }
    override fun add(element: SmartDevice): Boolean {
        val success = super.add(element)
        if (success) {
            listeners.removeIf { it.get() == null }
            listeners.forEach { it.get()?.invoke() }
        }
        return success
    }
    fun onChange(call: ()->Unit) = listeners.add(WeakReference(call))
}

typealias WeakListeners = ArrayList<WeakReference<()->Unit>>

